Feature/user-signup

nayoonk928 opened the pull request:
### 변경사항
**AS-IS**

- 사용자 회원 가입 구현
  - UserController에 userSignUp을 구현하여 사용자 등록에 대한 POST 요청 처리
  - UserServiceImpl을 추가하여 사용자 등록 로직 처리
  - 중복된 사용자 ID, 이메일, 닉네임을 확인하는 유효성 검사 구현
  - BCryptPasswordEncoder를 사용하여 사용자 비밀번호 암호화
  - UserRepository를 사용하여 새로운 사용자 데이터베이스에 저장
  - UserController 및 UserServiceImpl 클래스에 대한 테스트 진행
  
- 예외 처리
  - CustomException 클래스를 생성하여 사용자 정의 예외 처리
  - ErrorCode enum 클래스를 추가하여 다양한 예외 유형 관리
  - GlobalExceptionHandler 클래스를 생성하여 애플리케이션 전역 예외 처리 구현
  - GlobalExceptionHandler에서 CusteomException을 캐치하여 적절한 HTTP 상태 코드와 메시지 반환

**TO-BE**

- 로그인, JWT 토큰 발행
- 초기 관리자 생성

### 테스트
- [x] 테스트 
- [x] API 테스트

nayoonk928 committed:
chore: change application.yml and build.gradle

nayoonk928 committed:
feat: Implement user sign-up feature

- UserController에 userSignUp을 구현하여 사용자 등록에 대한 POST 요청 처리
- UserServiceImpl을 추가하여 사용자 등록 로직 처리
- 중복된 사용자ID, 이메일, 닉네임을 확인하는 유효성 검사 구현
- BCryptPasswordEncoder를 사용하여 사용자 비밀번호 암호화
- UserRepository를 사용하여 새로운 사용자 데이터베이스에 저장
- UserController 및 UserServiceImpl 클래스에 대한 테스트 진행
- 예외처리를 위한 클래스 구현

mingkiii commented:
동시에 여러명의 사용자가 같은 아이디, 닉네임 등으로 회원가입을 요청할 때에 발생하는 '동시성 이슈'를 대비하는 부분이 없는 것 같아 추가 구현해보시면 좋을 것 같습니다. 비관적 락, 낙관적 락, redis락 등 여러 방법이 있습니다.

mingkiii commented on src/main/java/com/personal/contentservice/dto/SignUpDto.java:
```diff
@@ -0,0 +1,46 @@
+  private String userId;
+
+  @NotBlank(message = "이메일은 필수 항목 입니다.")
+  @EmailValidation()
```
커스텀어노테이션 대신 '@Email' 이메일 형식을 검사해주는 스프링에서 제공해주는 어노테이션이 있습니다.

nayoonk928 replied:
@Email 도 사용해봤지만 'newEmail@example' 형식도 올바른 형식으로 인식하더라구요. 혹시 @Email 어노테이션을 사용하면서 .com 과 같은 도메인 형식도 식별하는 방법이 있나요?

mingkiii replied:
그렇군요!그래서 구글링해보았는데요 '클라이언트로부터받은 양식 유효성 검사를 수행하거나 빈 유효성 검사를 수행하려는 경우 단순하게 유지하십시오. 웹서비스에 등록하려고 할 때 엄격하게 확인하고 일부 사람들을 거부하는 것보다 느슨한 확인을 수행하는 것이 좋습니다. 이메일의 사용자 이름부분에 거의 모든것이 허용되고 매월 말 추가되는 많은 새 도메인(예: .company, .entrtprise, .estate 등)이 제한적이지 않은 것이 더 안전합니다.'라는 글을 보았고 적용할 수 있는 방법으로는 
'@Pattern(regexp = "^.+@.+\\..+$")' 정규식으로 "문자+'@'+문자+'.'+문자"를 확인합니다.
덕분에 이메일 어노테이션이 도메인 형식을 검증하진 못한다는 사실을 알게 되었습니다. 저도 패턴 어노테이션으로 수정해야겠네요!

mingkiii commented on src/main/java/com/personal/contentservice/service/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,55 @@
+        .build();
+
+    userRepository.save(user);
+    return ResponseEntity.ok().body(user);
```
ResponseEntity에 user 객체를 그대로 내보내기보다 Dto로 내보내는게 좋다고 생각합니다.
(보안과 개인정보보호)Dto로 필요한 데이터만 담아서 클라이언트에게 노출할 수 있고, 민감한 정보나 불필요한 데이터가 노출 되지 않습니다. 그리고 도메인 모델과 표현 계층 간에 명확한 분리를 할 수 있어서 유지 보수성과 확장성이 더욱 높아집니다.

nayoonk928 replied:
감사합니다! response dto를 통해 비밀번호는 반환하지 않는 것으로 수정했습니다.

mingming-mentor commented on src/main/java/com/personal/contentservice/service/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,55 @@
+  // 회원 가입
+  @Override
+  @Transactional
+  public ResponseEntity<?> signUp(SignUpDto signUpDto) {
```
현재 구조에서도 회원가입이 단건씩 들어올땐 문제가 없지만, 여러명이 같은 email/nickname으로 회원가입을 요청하면 customException에 걸리지 않을 수 있을 듯해요. lock을 잡아서 email이나 nickname을 체크해주면 같은 이메일로 회원가입 요청시 의도한대로 customException으로 반환 가능할 것 같아요.

nayoonk928 replied:
Redisson Lock을 이용해서 수정했습니다! 감사합니다.

mingming-mentor commented on src/main/java/com/personal/contentservice/util/EmailValidator.java:
```diff
@@ -0,0 +1,24 @@
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+
+public class EmailValidator implements ConstraintValidator<EmailValidation, String> {
```
custom validator를 잘 사용하셨네요!

mingming-mentor commented on src/main/java/com/personal/contentservice/util/PasswordUtils.java:
```diff
@@ -0,0 +1,20 @@
+
+  // password 암호화해서 리턴하는 함수
+  public static String getEncryptPassword(String password) {
+    BCryptPasswordEncoder bCryptPasswordEncoder = new BCryptPasswordEncoder();
```
passwordEncoder의 경우 new해서 사용해주기보다는 bean으로 사용해주시면 좋을 것 같아요.
클래스가 무겁기때문에 매 요청마다 new를 통해 인스턴스를 만들어주면 비용이 많이 듭니다.

nayoonk928 replied:
bean 이용하는 것으로 수정 완료했습니다. 감사합니다.

mingming-mentor commented on src/main/java/com/personal/contentservice/domain/BaseEntity.java:
```diff
@@ -0,0 +1,42 @@
+  @LastModifiedDate
+  private LocalDateTime lastModifiedAt;
+
+  @PrePersist
```
@CreatedDate와 같은 역할을 하는 부분인데 중복되어 코드가 작성되어있는 듯해요.

nayoonk928 replied:
감사합니다. 중복 코드 삭제했습니다!

mingming-mentor reviewed (approved):
수고 많으셨습니다!

nayoonk928 committed:
chore: update build.gradle and application.yml

nayoonk928 committed:
refactor: remove userId field from User entity
- User 엔티티에서 userId 삭제
- SignUpDto 안에 Request 와 Response 생성
- signUp 메서드 반환 값을 SignUpDto.Response로 변경
- EmailValidation 정규식을 이용하도록 변경
- passwordEncoder를 new 대신 Bean으로 사용하도록 변경
- BaseEntity에서 createdAt, lastModifiedAt 관련 중복 코드 제거
- DataIntegrityViolationException 처리 추가

nayoonk928 committed:
feat: add redisson lock to sign-up functionality
- 회원가입 요청 간 동시성 문제를 방지하기 위해 Redisson lock 구현
- UserServiceImpl 업데이트하여 회원가입 시 Redisson lock 처리

nayoonk928 committed:
test: add and change test cases for sign-up and redisson lock

nayoonk928 committed:
refactor: change email regex
